board/safety/safety_nissan.h

```diff
@@ -0,0 +1,212 @@
+
+
+static int nissan_rx_hook(CAN_FIFOMailBox_TypeDef *to_push) {
+
+  bool valid = addr_safety_check(to_push, nissan_rx_checks, NISSAN_RX_CHECK_LEN,
+                                 NULL, NULL, NULL);
+
+  if (valid) {
+    int bus = GET_BUS(to_push);
+    int addr = GET_ADDR(to_push);
+
+    if (bus == 0) {
+      if (addr == 0x2) {
+        // Current steering angle
+        // Factor -0.1, little endian
+        int angle_meas_new = (GET_BYTES_04(to_push) & 0xFFFF);
+        // Need to multiply by 10 here as LKAS and Steering wheel are different base unit
+        angle_meas_new = to_signed(angle_meas_new, 16) * 10;
+
+        // update array of samples
+        update_sample(&nissan_angle_meas, angle_meas_new);
+      }
+
+      if (addr == 0x29a) {
+        // Get current speed
+        // Factor 0.00555
+        nissan_speed = ((GET_BYTE(to_push, 2) << 8) | (GET_BYTE(to_push, 3))) * 0.00555 / 3.6;
+      }
+
+      // exit controls on rising edge of gas press
+      if (addr == 0x15c) {
+        int gas = ((GET_BYTE(to_push, 5) << 2) | ((GET_BYTE(to_push, 6) >> 6) & 0x3));
+        if ((gas > 0) && (nissan_gas_prev == 0)) {
+          controls_allowed = 0;
+        }
+        nissan_gas_prev = gas;
+      }
+
+      // 0x169 is lkas cmd. If it is on bus 0, then relay is unexpectedly closed
+      if ((safety_mode_cnt > RELAY_TRNS_TIMEOUT) && (addr == 0x169)) {
+        relay_malfunction = true;
+      }
+    }
+
+    if (bus == 1) {
+      if (addr == 0x1b6) {
+        int cruise_engaged = (GET_BYTE(to_push, 4) >> 6) & 1;
+        if (cruise_engaged && !nissan_cruise_engaged_last) {
+          controls_allowed = 1;
+        }
+        if (!cruise_engaged) {
+          controls_allowed = 0;
+        }
+        nissan_cruise_engaged_last = cruise_engaged;
+      }
+
+      // exit controls on rising edge of brake press if speed > 0
+      if (addr == 0x454) {
+        int brake = (GET_BYTE(to_push, 2) & 0x80);
+        if ((brake > 0) && (nissan_brake_prev == 0) && (nissan_speed > 0.)) {
+          controls_allowed = 0;
+        }
+        nissan_brake_prev = brake;
+      }
+    }
+  }
+  return valid;
+}
+
+
+static int nissan_tx_hook(CAN_FIFOMailBox_TypeDef *to_send) {
+  int tx = 1;
+  int addr = GET_ADDR(to_send);
+  int bus = GET_BUS(to_send);
+
+  if (!msg_allowed(addr, bus, NISSAN_TX_MSGS, sizeof(NISSAN_TX_MSGS) / sizeof(NISSAN_TX_MSGS[0]))) {
+    tx = 0;
+  }
+
+  if (relay_malfunction) {
+    tx = 0;
+  }
+
+  // steer cmd checks
+  if (addr == 0x169) {
+    int desired_angle = ((GET_BYTE(to_send, 0) << 10) | (GET_BYTE(to_send, 1) << 2) | ((GET_BYTE(to_send, 2) >> 6) & 0x3));
+    bool lka_active = (GET_BYTE(to_send, 6) >> 4) & 1;
+
+    // offeset 1310 * NISSAN_DEG_TO_CAN
+    desired_angle =  desired_angle - 131000;
+    bool violation = 0;
+
+    if (controls_allowed && lka_active) {
+      // add 1 to not false trigger the violation
+      float delta_angle_float;
```

Comment:
Not for now, but this code is the same as the one used in other safety files (toyota_ipas and maybe Tesla?), so it should be abstracted. I'll look into this.